### PR TITLE
Extend E2E test to include coverage for creating Operator from OperatorSource

### PR DESCRIPTION
**Pull request by ldimaggi**
This pull request will include:
* OperatorSource definition file extend E2E test to include verifying creating Operator from OperatorSource as defined here: https://github.com/operator-framework/operator-marketplace
* Updates to the E2E test

Ref: https://jira.coreos.com/browse/ODC-580

### CODE REVIEW

File: test/e2e/devconsole.operatorsource.cr.yaml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+apiVersion: operators.coreos.com/v1
+kind: OperatorSource
+metadata:
+  name: devconsole-operator
+  namespace: openshift-operators
+spec:
+  type: appregistry
+  endpoint: https://quay.io/repository/redhat-developer/devconsole-operator
```
Comment (sbose78) [inline test/e2e/devconsole.operatorsource.cr.yaml]: Does this actually work like an app registry?

Comment (ldimaggi) [reply test/e2e/devconsole.operatorsource.cr.yaml]: It allowed the subscription to be created - but you are correct - it's not an app registry. Will retry after having created an app registry in Quay.

File: test/e2e/devconsole.operatorsource.cr.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+apiVersion: operators.coreos.com/v1
+kind: OperatorSource
+metadata:
+  name: devconsole-operator
+  namespace: openshift-operators
+spec:
+  type: appregistry
+  endpoint: https://quay.io/application/ldimaggi-test-org/mychart
```
Comment (sbose78) [inline test/e2e/devconsole.operatorsource.cr.yaml]: Nice! Do paste a screenshot in this PR how it looks in the operatorhub page on the cluster.

Comment (ldimaggi) [reply test/e2e/devconsole.operatorsource.cr.yaml]: I spent a lot of time with Helm this morning and I *think* I have the correct sequence - but while the subscription is showing up inthe UI, the catalog is not. The problem that I am seeing is:  https://github.com/operator-framework/operator-lifecycle-manager/issues/822 - currently blocked on this.

**Comment by sbose78** [comment]
What are the `oc` commands you've tried to debug?

**Comment by ldimaggi** [comment]
> What are the `oc` commands you've tried to debug?

kubectl apply -f devconsole.operatorsource.catalog.yaml

**Comment by pmacik** [comment]
@ldimaggi The resource that is being created is `CatalogSourceConfig` with a name of `devconsole-operator-catalog` - [here](https://github.com/ldimaggi/devconsole-operator/blob/580-operator-source/test/e2e/devconsole.operatorsource.catalog.yaml#L4).
But in the `Subscription` where you try to reference it by `spec.source: devconsole-operator-catalog` - [here](https://github.com/ldimaggi/devconsole-operator/blob/580-operator-source/test/e2e/devconsole.operatorsource.subscription.yaml#L9) - what the subscription is actually referencing is a `CatalogSource` instance, not `CatalogSourceConfig`.

@sbose78 WDYT, do I understand it correctly?

**Comment by ldimaggi** [comment]
> @ldimaggi The resource that is being created is `CatalogSourceConfig` with a name of `devconsole-operator-catalog` - [here](https://github.com/ldimaggi/devconsole-operator/blob/580-operator-source/test/e2e/devconsole.operatorsource.catalog.yaml#L4).
> But in the `Subscription` where you try to reference it by `spec.source: devconsole-operator-catalog` - [here](https://github.com/ldimaggi/devconsole-operator/blob/580-operator-source/test/e2e/devconsole.operatorsource.subscription.yaml#L9) - what the subscription actually references is a `CatalogSource` instance (and that does not exist), not `CatalogSourceConfig`.
> 
> @sbose78 WDYT, do I understand it correctly?

I modeled this on these files:

https://github.com/redhat-developer/devconsole-operator/blob/master/test/e2e/catalog_source_OS4.yaml

https://github.com/redhat-developer/devconsole-operator/blob/master/test/e2e/subscription_OS4.yaml

**Comment by ldimaggi** [comment]
It seems like this is a contributing factor to the subscription failing:

Create a CatalogSource:
```
cat devconsole.operatorsource.catalog.yamlapiVersion: operators.coreos.com/v1
kind: CatalogSourceConfig
metadata:
  name: devconsole-operator-catalog
  namespace: openshift-operators
spec:
  targetNamespace: local-operators
  packages: devconsole-operator
  displayName: Red Hat Devconsole Operator
  publisher: Red Hat
```

Apply the CatalogSource:
```
kubectl apply -f devconsole.operatorsource.catalog.yamlcatalogsourceconfig "devconsole-operator-catalog" configured
```

Verify the CatalogSource in the CLI:
```
kubectl get CatalogSourceConfig devconsole-operator-catalog -n openshift-operators -o=yaml
apiVersion: operators.coreos.com/v1
kind: CatalogSourceConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"operators.coreos.com/v1","kind":"CatalogSourceConfig","metadata":{"annotations":{},"name":"devconsole-operator-catalog","namespace":"openshift-operators"},"spec":{"displayName":"Red Hat Devconsole Operator","packages":"devconsole-operator","publisher":"Red Hat","targetNamespace":"local-operators"}}
  creationTimestamp: 2019-04-25T17:19:11Z
  generation: 1
  name: devconsole-operator-catalog
  namespace: openshift-operators
  resourceVersion: "41007"
  selfLink: /apis/operators.coreos.com/v1/namespaces/openshift-operators/catalogsourceconfigs/devconsole-operator-catalog
  uid: 3e1a55ff-677e-11e9-8124-06472ef5fc02
spec:
  displayName: Red Hat Devconsole Operator
  packages: devconsole-operator
  publisher: Red Hat
  targetNamespace: local-operators
```

But - the CatalogSource is not displayed in the UI:
Screenshot from 2019-04-25 13-34-39

**Comment by ldimaggi** [comment]
See:
https://github.com/operator-framework/operator-lifecycle-manager/issues/822

File: test/e2e/devconsole.operatorsource.cr.yaml (hunk 2 above)
Comment (pmacik) [inline test/e2e/devconsole.operatorsource.cr.yaml]: According to the [official example](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/examples/community.operatorsource.cr.yaml) the 
[`spec.endpoint`](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/examples/community.operatorsource.cr.yaml#L5) should be `https://quay.io/cnr`
and [`spec.registryNamespace`](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/examples/community.operatorsource.cr.yaml#L8) should be `ldimaggi-test-org`

File: test/e2e/devconsole.operatorsource.catalog.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+apiVersion: operators.coreos.com/v1
+kind: CatalogSourceConfig
+metadata:
+  name: devconsole-operator-catalog
+  namespace: openshift-operators
```
Comment (pmacik) [inline test/e2e/devconsole.operatorsource.catalog.yaml]: According to the [official example](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/examples/catalogsourceconfig.cr.yaml):

* [`metadata.namespace`](https://github.com/operator-framework/operator-marketplace/blob/master/deploy/examples/catalogsourceconfig.cr.yaml#L5) should be `openshift-marketplace`

**Comment by pmacik** [comment]
I just realized, that the `OperatorSource` or `CatalogSourceConfig` should probably be placed in the `operatoror-marketplace` namespace:
```sh
(operator-test-env)➜   oc get csc --all-namespaces
NAMESPACE               NAME                  STATUS      MESSAGE                                       AGE
openshift-marketplace   certified-operators   Succeeded   The object has been successfully reconciled   29h
openshift-marketplace   community-operators   Succeeded   The object has been successfully reconciled   29h
openshift-marketplace   redhat-operators      Succeeded   The object has been successfully reconciled   29h

(operator-test-env)➜   oc get catsrc --all-namespaces
NAMESPACE                              NAME                  NAME                  TYPE       PUBLISHER   AGE
openshift-marketplace                  certified-operators   Certified Operators   grpc       Red Hat     29h
openshift-marketplace                  community-operators   Community Operators   grpc       Red Hat     29h
openshift-marketplace                  redhat-operators      Red Hat Operators     grpc       Red Hat     29h
openshift-operator-lifecycle-manager   olm-operators         OLM Operators         internal   Red Hat     30h

(operator-test-env)➜   oc get opsrc --all-namespaces
NAMESPACE               NAME                  TYPE          ENDPOINT              REGISTRY              DISPLAYNAME           PUBLISHER   STATUS      MESSAGE                                       AGE
openshift-marketplace   certified-operators   appregistry   https://quay.io/cnr   certified-operators   Certified Operators   Red Hat     Succeeded   The object has been successfully reconciled   30h
openshift-marketplace   community-operators   appregistry   https://quay.io/cnr   community-operators   Community Operators   Red Hat     Succeeded   The object has been successfully reconciled   30h
openshift-marketplace   redhat-operators      appregistry   https://quay.io/cnr   redhat-operators      Red Hat Operators     Red Hat     Succeeded   The object has been successfully reconciled   30h

```

And when an `OperatorSource` is created in that namespace, the `CatalogSourceConfig` and `CatalogSource` are created right away with the same name:
```sh
(operator-test-env)➜  mychart cat devconsole.operatorsource.yaml 
```
```yaml
apiVersion: operators.coreos.com/v1
kind: OperatorSource
metadata:
  name: devconsole-operators
  namespace: openshift-marketplace
spec:
  type: appregistry
  endpoint: https://quay.io/cnr
  registryNamespace: odcqe
  displayName: "Red Hat Devconsole Operator"
  publisher: "Red Hat"
```
```sh
(operator-test-env)➜  oc get csc --all-namespaces          
NAMESPACE               NAME                   STATUS      MESSAGE                                       AGE
openshift-marketplace   certified-operators    Succeeded   The object has been successfully reconciled   30h
openshift-marketplace   community-operators    Succeeded   The object has been successfully reconciled   30h
openshift-marketplace   devconsole-operators   Succeeded   The object has been successfully reconciled   90s
openshift-marketplace   redhat-operators       Succeeded   The object has been successfully reconciled   30h

(operator-test-env)➜  oc get catsrc --all-namespaces 
NAMESPACE                              NAME                   NAME                          TYPE       PUBLISHER   AGE
openshift-marketplace                  certified-operators    Certified Operators           grpc       Red Hat     30h
openshift-marketplace                  community-operators    Community Operators           grpc       Red Hat     30h
openshift-marketplace                  devconsole-operators   Red Hat Devconsole Operator   grpc       Red Hat     98s
openshift-marketplace                  redhat-operators       Red Hat Operators             grpc       Red Hat     30h
openshift-operator-lifecycle-manager   olm-operators          OLM Operators                 internal   Red Hat     30h

(operator-test-env)➜  oc get opsrc --all-namespaces
NAMESPACE               NAME                   TYPE          ENDPOINT              REGISTRY              DISPLAYNAME                   PUBLISHER   STATUS      MESSAGE                                       AGE
openshift-marketplace   certified-operators    appregistry   https://quay.io/cnr   certified-operators   Certified Operators           Red Hat     Succeeded   The object has been successfully reconciled   30h
openshift-marketplace   community-operators    appregistry   https://quay.io/cnr   community-operators   Community Operators           Red Hat     Succeeded   The object has been successfully reconciled   30h
openshift-marketplace   devconsole-operators   appregistry   https://quay.io/cnr   odcqe                 Red Hat Devconsole Operator   Red Hat     Succeeded   The object has been successfully reconciled   102s
openshift-marketplace   redhat-operators       appregistry   https://quay.io/cnr   redhat-operators      Red Hat Operators             Red Hat     Succeeded   The object has been successfully reconciled   30h
```

**Comment by pmacik** [comment]
Then a pod is created:
```sh
(operator-test-env)➜  oc get pods -n openshift-marketplace

NAME                                    READY   STATUS             RESTARTS   AGE
certified-operators-d58db9c6b-thm2r     1/1     Running            0          30h
community-operators-7cfbbbb6b-v5cvg     1/1     Running            0          30h
devconsole-operators-585dcb889c-4lv2p   0/1     CrashLoopBackOff   9          22m
marketplace-operator-5cc4974ddc-mqzmt   1/1     Running            0          30h
redhat-operators-77848897d7-s28vh       1/1     Running            0          30h
```
but the pod fails:
```sh
(operator-test-env)➜  ~ oc logs devconsole-operators-585dcb889c-4lv2p -n openshift-marketplace
time="2019-04-26T13:12:16Z" level=info msg="Using in-cluster kube client config" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="operator source(s) specified are - [https://quay.io/cnr|odcqe]" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="package(s) specified are - devconsole-operator-test" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="input has been sanitized" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="sources: [https://quay.io/cnr/odcqe]" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="packages: [devconsole-operator-test]" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="resolved the following packages: [odcqe/devconsole-operator-test:0.1.0]" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="downloading repository: odcqe/devconsole-operator-test:0.1.0 from https://quay.io/cnr" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="download complete - 1 repositories have been downloaded" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="decoding the downloaded operator manifest(s)" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="manifest format is - nested" port=50051 repository="odcqe/devconsole-operator-test:0.1.0" type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="downloaded/devconsole-operator-test/devconsole.operatorsource.catalog.yaml - type=file" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=error msg="skipping due to error - error happened while processing tar file - open downloaded/devconsole-operator-test/devconsole.operatorsource.catalog.yaml: no such file or directory" port=50051 repository="odcqe/devconsole-operator-test:0.1.0" type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="decoded 0 flattened and 1 nested operator manifest(s)" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=info msg="loading nested operator bundle(s) from downloaded into sqlite" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=error msg="stat failed on target directory[downloaded] - stat downloaded: no such file or directory" port=50051 type=appregistry
time="2019-04-26T13:12:16Z" level=fatal msg="error loading manifest from remote registry - stat downloaded: no such file or directory" port=50051 type=appregistry
```

**Comment by pmacik** [comment]
This line in the pod log got my attention:
`time="2019-04-26T13:12:16Z" level=error msg="skipping due to error - error happened while processing tar file - open downloaded/devconsole-operator-test/devconsole.operatorsource.catalog.yaml: no such file or directory" port=50051 repository="odcqe/devconsole-operator-test:0.1.0" type=appregistry`

**Comment by pmacik** [comment]
@ldimaggi We should split the `OperatorSource` test to 2 steps - first is pushing the operator into the app registry and then installing the operator via the `OperatorSource` from that app registry.

https://github.com/redhat-developer/devconsole-operator/pull/174 Covers the pushing of the operator to the Quay app registry.

**Comment by pmacik** [comment]
This PR is made obsolete by #181

**Comment by ldimaggi** [comment]
Closing this pull request - un-merged - in favor of pull request #181

**Pull request closed without merging**